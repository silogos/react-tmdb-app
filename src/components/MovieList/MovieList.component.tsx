import { DisplayError, DisplayNoData } from "@/components/Displays";

function MovieCardShimmenring() {
  return (
    <div className="animate-pulse aspect-[8/12] rounded-lg bg-gray-700 bg-opacity-90" />
  );
}

export function ListLoading() {
  return Array(20)
    .fill(true)
    .map((_, idx) => <MovieCardShimmenring key={idx.toString()} />);
}

export function ListError() {
  return <DisplayError />;
}

export function ListEmpty() {
  return <DisplayNoData />;
}
