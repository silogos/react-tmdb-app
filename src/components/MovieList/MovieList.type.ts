import { Movie } from "@/types/Movie.types";

export type MovieListProps = {
  movies: Movie[];
  isLoading: boolean;
  isError: boolean;
  onReload?: () => void;
  hasMoreData: boolean;
  onLoadMore: () => void;
};

export type MoviListErrorProps = {
  onReload?: () => void;
};
