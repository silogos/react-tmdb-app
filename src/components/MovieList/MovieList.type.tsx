import { Movie } from "@/types/Movie.types";

export type MovieListProps = {
  movies: Movie[];
  isLoading: boolean;
  isError: boolean;
  hasMoreData: boolean;
  onLoadMore: () => void;
};
