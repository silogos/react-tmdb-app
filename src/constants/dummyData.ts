import { Credit, Movie, MovieDetail } from "@/types/Movie.types";

export const dummyMovies: Movie[] = [
  {
    adult: false,
    backdrop_path: "/b85bJfrTOSJ7M5Ox0yp4lxIxdG1.jpg",
    genre_ids: [28, 878, 35, 10751],
    id: 939243,
    original_language: "en",
    original_title: "Sonic the Hedgehog 3",
    overview:
      "Sonic, Knuckles, and Tails reunite against a powerful new adversary, Shadow, a mysterious villain with powers unlike anything they have faced before. With their abilities outmatched in every way, Team Sonic must seek out an unlikely alliance in hopes of stopping Shadow and protecting the planet.",
    popularity: 7535.831,
    poster_path: "/d8Ryb8AunYAuycVKDp5HpdWPKgC.jpg",
    release_date: "2024-12-19",
    title: "Sonic the Hedgehog 3",
    video: false,
    vote_average: 7.8,
    vote_count: 1078,
  },
  {
    adult: false,
    backdrop_path: "/xZm5YUNY3PlYD1Q4k7X8zd2V4AK.jpg",
    genre_ids: [28, 35],
    id: 993710,
    original_language: "en",
    original_title: "Back in Action",
    overview:
      "Fifteen years after vanishing from the CIA to start a family, elite spies Matt and Emily jump back into the world of espionage when their cover is blown.",
    popularity: 3559.004,
    poster_path: "/3L3l6LsiLGHkTG4RFB2aBA6BttB.jpg",
    release_date: "2025-01-15",
    title: "Back in Action",
    video: false,
    vote_average: 6.724,
    vote_count: 533,
  },
];

export const dummyMovie: Movie = dummyMovies[0];

export const dummyMovieDetail: MovieDetail = {
  adult: false,
  backdrop_path: "/b85bJfrTOSJ7M5Ox0yp4lxIxdG1.jpg",
  belongs_to_collection: {
    id: 720879,
    name: "Sonic the Hedgehog Collection",
    poster_path: "/fwFWhYXj8wY6gFACtecJbg229FI.jpg",
    backdrop_path: "/l5CIAdxVhhaUD3DaS4lP4AR2so9.jpg",
  },
  budget: 122000000,
  genres: [
    {
      id: 28,
      name: "Action",
    },
    {
      id: 878,
      name: "Science Fiction",
    },
    {
      id: 35,
      name: "Comedy",
    },
    {
      id: 10751,
      name: "Family",
    },
  ],
  homepage: "https://www.sonicthehedgehogmovie.com",
  id: 939243,
  imdb_id: "tt18259086",
  origin_country: ["US"],
  original_language: "en",
  original_title: "Sonic the Hedgehog 3",
  overview:
    "Sonic, Knuckles, and Tails reunite against a powerful new adversary, Shadow, a mysterious villain with powers unlike anything they have faced before. With their abilities outmatched in every way, Team Sonic must seek out an unlikely alliance in hopes of stopping Shadow and protecting the planet.",
  popularity: 3560.42,
  poster_path: "/d8Ryb8AunYAuycVKDp5HpdWPKgC.jpg",
  production_companies: [
    {
      id: 4,
      logo_path: "/gz66EfNoYPqHTYI4q9UEN4CbHRc.png",
      name: "Paramount Pictures",
      origin_country: "US",
    },
    {
      id: 333,
      logo_path: "/5xUJfzPZ8jWJUDzYtIeuPO4qPIa.png",
      name: "Original Film",
      origin_country: "US",
    },
    {
      id: 77884,
      logo_path: "/dP2lxVNctD5Cried0IWVqgrO2o9.png",
      name: "Marza Animation Planet",
      origin_country: "JP",
    },
    {
      id: 113750,
      logo_path: "/A3QVZ9Ah0yI2d2GiXUFpdlbTgyr.png",
      name: "SEGA",
      origin_country: "JP",
    },
    {
      id: 10644,
      logo_path: "/ocLZIdYJBppuCt1rhYEb2jbpt5F.png",
      name: "Blur Studio",
      origin_country: "US",
    },
    {
      id: 168701,
      logo_path: "/vWdZFT4V64CCv12D10m44duQjyg.png",
      name: "SEGA of America",
      origin_country: "US",
    },
  ],
  production_countries: [
    {
      iso_3166_1: "JP",
      name: "Japan",
    },
    {
      iso_3166_1: "US",
      name: "United States of America",
    },
  ],
  release_date: "2024-12-19",
  revenue: 422398000,
  runtime: 110,
  spoken_languages: [
    {
      english_name: "English",
      iso_639_1: "en",
      name: "English",
    },
    {
      english_name: "French",
      iso_639_1: "fr",
      name: "Français",
    },
    {
      english_name: "German",
      iso_639_1: "de",
      name: "Deutsch",
    },
    {
      english_name: "Japanese",
      iso_639_1: "ja",
      name: "日本語",
    },
    {
      english_name: "Letzeburgesch",
      iso_639_1: "lb",
      name: "",
    },
    {
      english_name: "Malagasy",
      iso_639_1: "mg",
      name: "",
    },
    {
      english_name: "Spanish",
      iso_639_1: "es",
      name: "Español",
    },
  ],
  status: "Released",
  tagline: "Try to keep up.",
  title: "Sonic the Hedgehog 3",
  video: false,
  vote_average: 7.597,
  vote_count: 725,
};

export const dummyCredits: Credit[] = [
  {
    adult: false,
    gender: 2,
    id: 27428,
    known_for_department: "Acting",
    name: "Aaron Taylor-Johnson",
    original_name: "Aaron Taylor-Johnson",
    popularity: 49.234,
    profile_path: "/pFtHhih2XEaFaD3qOFyQW6q83br.jpg",
    cast_id: 8,
    character: "Sergei Kravinoff / Kraven the Hunter",
    credit_id: "60aec3f9d29bdd002c022ce0",
    order: 0,
  },
  {
    adult: false,
    gender: 1,
    id: 1437491,
    known_for_department: "Acting",
    name: "Ariana DeBose",
    original_name: "Ariana DeBose",
    popularity: 17.829,
    profile_path: "/hEamSutadgLbVNHch3KhJ5LfPdp.jpg",
    cast_id: 14,
    character: "Calypso Ezili",
    credit_id: "621e8519f12cf4001b7f2ccf",
    order: 1,
  },
];
export const dummyCredit: Credit = dummyCredits[0];
