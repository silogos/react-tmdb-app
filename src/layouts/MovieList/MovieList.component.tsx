export function Hero() {
  return (
    <section id="hero" className="overflow-hidden py-4">
      <div className="mx-auto max-w-screen-xl px-4 py-14 lg:px-8 lg:py-20">
        <h1 className="text-center text-gray-300 text-4xl lg:text-7xl mb-4">
          ðŸŽ¬ Movie App
        </h1>
        <p className="text-center text-gray-300 text-base lg:text-2xl">
          (The Movie Database)
        </p>
      </div>
    </section>
  );
}
