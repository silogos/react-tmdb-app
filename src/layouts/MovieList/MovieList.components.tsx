import { useNavigate } from "react-router";

export function MovieListHero() {
  return (
    <section id="hero" className="overflow-hidden py-4">
      <div className="mx-auto max-w-screen-xl px-4 py-14 lg:px-8 lg:py-20">
        <h1 className="text-center text-gray-300 text-4xl lg:text-7xl mb-4">
          TMDB
        </h1>
        <p className="text-center text-gray-300 text-base lg:text-2xl">
          (The Movie Database)
        </p>
      </div>
    </section>
  );
}

export function SearchInput() {
  const navigate = useNavigate();

  return (
    <form
      onSubmit={(evt) => {
        evt.preventDefault();
        console.log("Submited");
      }}
    >
      <input
        type="search"
        className="w-64 border-2 rounded-lg  px-4 py-2 bg-gray-300"
        placeholder="Search..."
        onKeyDown={(evt) => {
          if (evt.currentTarget.checkValidity() && evt.key === "Enter") {
            console.log("do validate", !!evt.currentTarget.value);
            navigate(`/movies/search?query=${evt.currentTarget.value}`);
          }
        }}
        minLength={3}
        required
      />
    </form>
  );
}
