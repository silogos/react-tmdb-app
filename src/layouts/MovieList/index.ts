export { default } from "./MovieList.layout";
