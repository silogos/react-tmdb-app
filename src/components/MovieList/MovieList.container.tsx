import MovieCard from "@/components/MovieCard";
import { MovieListProps } from "./MovieList.type";
import { ListEmpty, ListError, ListLoading } from "./MovieList.component";

function MovieList({
  movies,
  isLoading,
  isError,
  hasMoreData,
  onLoadMore,
}: MovieListProps) {
  return (
    <>
      <div className="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4">
        {movies.map((movie) => (
          <MovieCard
            key={movie.id}
            id={movie.id}
            poster={movie.poster_path}
            releaseDate={movie.release_date}
            title={movie.title}
            voteAverage={movie.vote_average}
          />
        ))}
        {isLoading && <ListLoading />}
      </div>

      {!isLoading && (
        <>
          {isError && <ListError />}
          {!isError && movies.length < 1 && <ListEmpty />}
          {hasMoreData && (
            <button
              onClick={onLoadMore}
              className="my-8 w-full text-center py-2 p-8 bg-gray-300 rounded-lg"
            >
              Load More
            </button>
          )}
        </>
      )}
    </>
  );
}

export default MovieList;
