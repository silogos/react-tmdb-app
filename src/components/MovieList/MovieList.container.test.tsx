import { describe, it, expect, vitest, vi } from "vitest";
import MovieList from "./MovieList.container";
import { fireEvent, render } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import { dummyMovies } from "@/constants/dummyData";

vi.mock("@/components/MovieCard/MovieCard.component.tsx");
vi.mock("./MovieList.component");

describe("MovieList.component", () => {
  it("renders correctly with data", () => {
    const component = render(
      <BrowserRouter>
        <MovieList
          movies={dummyMovies}
          isLoading={false}
          isError={false}
          hasMoreData={false}
          onLoadMore={() => {}}
        />
      </BrowserRouter>
    );

    expect(component.getAllByTestId("MovieCard")).toHaveLength(2);
    expect(component.container).toMatchSnapshot();
  });

  it("renders correctly with nodata state", () => {
    const component = render(
      <MovieList
        movies={[]}
        isLoading={false}
        isError={false}
        hasMoreData={false}
        onLoadMore={() => {}}
      />
    );

    expect(component.getByTestId("ListEmpty")).toBeInTheDocument();
    expect(component.container).toMatchSnapshot();
  });

  it("renders correctly when isLoading true", () => {
    const component = render(
      <MovieList
        movies={[]}
        isLoading={true}
        isError={false}
        hasMoreData={false}
        onLoadMore={() => {}}
      />
    );

    expect(component.getByTestId("ListLoading")).toBeInTheDocument();
    expect(component.container).toMatchSnapshot();
  });

  it("renders correctly when isError true", () => {
    const component = render(
      <MovieList
        movies={[]}
        isLoading={false}
        isError={true}
        hasMoreData={false}
        onLoadMore={() => {}}
      />
    );

    expect(component.getByTestId("ListError")).toBeInTheDocument();
    expect(component.container).toMatchSnapshot();
  });

  it("renders correctly when hasMoreData true", () => {
    const component = render(
      <MovieList
        movies={[]}
        isLoading={false}
        isError={false}
        hasMoreData={true}
        onLoadMore={() => {}}
      />
    );

    expect(component.getByText("Load More")).toBeInTheDocument();
    expect(component.container).toMatchSnapshot();
  });

  it("renders correctly when hasMoreData true and click it", () => {
    const onLoadMore = vitest.fn();
    const component = render(
      <MovieList
        movies={[]}
        isLoading={false}
        isError={false}
        hasMoreData={true}
        onLoadMore={onLoadMore}
      />
    );
    expect(component).toBeTruthy();
    expect(component.getByText("Load More")).toBeInTheDocument();

    const loadMoreButton = component.getByRole("button");
    fireEvent.click(loadMoreButton);
    expect(onLoadMore).toHaveBeenCalledOnce();
  });
});
