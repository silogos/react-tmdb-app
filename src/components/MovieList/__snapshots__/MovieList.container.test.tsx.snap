// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieList.component > renders correctly when hasMoreData true 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4"
  />
  <div
    data-testid="ListEmpty"
  />
  <button
    class="my-8 w-full text-center py-2 p-8 bg-gray-300 rounded-lg"
  >
    Load More
  </button>
</div>
`;

exports[`MovieList.component > renders correctly when isError true 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4"
  />
  <div
    data-testid="ListError"
  />
</div>
`;

exports[`MovieList.component > renders correctly when isLoading true 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4"
  >
    <div
      data-testid="ListLoading"
    />
  </div>
</div>
`;

exports[`MovieList.component > renders correctly with data 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4"
  >
    <div
      data-testid="MovieCard"
    />
    <div
      data-testid="MovieCard"
    />
  </div>
</div>
`;

exports[`MovieList.component > renders correctly with nodata state 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4"
  />
  <div
    data-testid="ListEmpty"
  />
</div>
`;
